Ergonomic format_args!

xen0n opened the pull request:
Fixes #9456 (at last).

Not a ground-up rewrite of the existing machinery, but more like an added intermediary layer between macro arguments and format placeholders. This is now implementing Rust RFC 1618!

rust-highfive commented:
r? @aturon

(rust_highfive has picked a reviewer for you, use r? to override)

xen0n commented:
This caused some small regressions here and there. I'll be slowly fixing them; compilation time is very long.

Non-regressions, just TODO:

* [ ] `compile-fail/ifmt-bad-arg.rs`
* [ ] `run-pass/ifmt.rs`

Noticed regressions:

* [ ] `compile-fail/macro-backtrace-println.rs`
* [ ] doctest `fmt_3` (`{:.*}` didn't eat the precision argument)

xen0n commented:
Implemented correct implicit args handling by resolving all `CountIsParam(i)`s and `ArgumentNext`s into absolute references during parse. Which means, for example, `{foo:.*} {} {1:.*} {:.*} {bar:.*}` now parses exactly like `{foo:.0$} {1} {1:.2$} {4:.3$} {bar:.5$}`. Which obviates the need to handle any next-references during expansion and outputting; will remove these later, after I finish up an RFC for this PR later this week.

aturon commented:
r? @alexcrichton

alexcrichton commented:
Thanks for the PR! I'll take a closer look at this once https://github.com/rust-lang/rfcs/pull/1618 is handled

xen0n commented:
Fixed test cases broken by this PR, and added new cases as well.

Currently `make check` fails on several unrelated tests:

* `run-pass/issue-28950.rs`
* `run-pass/panic-runtime/lto-abort.rs`
* `run-pass/panic-runtime/lto-unwind.rs`
* `run-pass/sepcomp-lib-lto.rs`

I'm not sure why those keep popping up, it wasn't the case several days ago. Testing master branch right now, will edit this once finished.

xen0n commented:
Travis test completed without the unrelated failures, seems there is a pretty-print round-trip failure:

```diff
--- pretty.expected.rs	2016-05-22 13:39:14.080831211 +0800
+++ pretty.actual.rs	2016-05-22 13:38:56.912831929 +0800
@@ -178,7 +178,6 @@
        , b = "hijklmn" , c = 3 ) , "abcd hijk 4\nabc hij 3");
     t!(format ! ( "{a:.*} {0} {:.*}" , 4 , 3 , "efgh" , a = "abcdef" ) ,
        "abcd 4 efg");
-
     // Test that pointers don't get truncated.
     {
         let val = usize::MAX;
```

Apart from this and deduplication of argument objects, the feature feels almost ready. Waiting for RFC resolution :smile:

alexcrichton commented:
Ah yeah unfortunately we've had bad luck with debuginfo and LTO in the past (lots of weird LLVM assertions), but thanks for the progress update!

xen0n commented:
I immediately gave up fixing the pretty print issue after trying to produce a minimal test case and got this:

```rust
macro_rules! t {
    ($a:expr, $b:expr) => { assert_eq!($a, $b) }
}

pub fn main() {
    t!(format!("{:x}{:X}{:x}{:X}{:x}{a:X}",1,2,3,4,5,a=15), "abcdefghijkl");




    // every time this code get pretty-printed, a newline is eaten!
}
```

Might as well reduce the length of affected lines, as the bug seems to only occur when a line wrap is needed in the output. I'll finish the argument object deduplication and then we can start reviewing.

xen0n commented:
Rebased and cleaned up a little, should be ready for review!


PS: The reason I think an RFC is really needed, is that some serious documentation changes and maybe deprecation inside the formatting runtime will have to be done. However as the RFC is already submitted and waiting through its FCP this point is already moot.

alexcrichton commented on src/test/run-pass/ifmt.rs:
```diff
@@ -163,6 +163,28 @@ pub fn main() {
+    // named arg only used as count
+    t!(format!("{0:.a$}", "aaaaaa", a=2), "aa");
+
+
```
Can you add a test here as well with side effects on the evaluated expression to ensure it's only calculated once even though it may be formatted multiple times.

alexcrichton commented on src/libsyntax_ext/format.rs:
```diff
@@ -626,7 +753,9 @@ pub fn expand_preparsed_format_args(ecx: &mut ExtCtxt, sp: Span,
-    let arg_types: Vec<_> = (0..args.len()).map(|_| None).collect();
+    // We can't use vec! here for Vec<ArgumentType>, as ArgumentType isn't Clone.
+    let arg_types: Vec<_> = (0..args.len()).map(|_| Vec::new()).collect();
+    let arg_unique_types: Vec<_> = (0..args.len()).map(|_| Vec::new()).collect();
```
I believe this can just be `vec![Vec::new(); args.len()]`

xen0n replied:
Nope, `ArgumentType` is not `Clone`, so `vec![something; num]`, which is `vec::from_elem` under the hood, will fail on this.

```rust
#[derive(PartialEq)]
enum ArgumentType {
    Placeholder(String),
    Count,
}

fn main() {
    let a: Vec<Vec<ArgumentType>> = vec![Vec::new(); 100];
}
```

No matter how you say `Vec::new()` this won't compile. Either we make `ArgumentType` `Clone` which I assume is not really desirable, or keep using the current way, which isn't that ugly after all.

alexcrichton commented on src/libfmt_macros/lib.rs:
```diff
@@ -77,9 +77,39 @@ pub struct FormatSpec<'a> {
+/// For internal use; same as FormatSpec except using CountInternal instead
+/// of Count. Possible next-references in the count fields will be resolved
+/// before the struct is converted into FormatSpec for returning.
+struct FormatSpecInternal<'a> {
```
Could we perhaps not add this and just handle all conversions and such in-place during parsing?

xen0n replied:
We don't have private enum variants at the moment, and I don't want to add a lot of `_` match arms to other places where the enums are consumed. But I'll look into this later though, currently outside.

alexcrichton commented on src/libsyntax_ext/format.rs:
```diff
@@ -46,13 +47,14 @@ struct Context<'a, 'b:'a> {
     /// Parsed argument expressions and the types that we've found so far for
     /// them.
```
Can you update this documentation for what `arg_types` and `arg_unique_types` are?

alexcrichton commented on src/libsyntax_ext/format.rs:
```diff
@@ -83,9 +98,9 @@ struct Context<'a, 'b:'a> {
               -> Option<(P<ast::Expr>, Vec<P<ast::Expr>>, Vec<String>,
                          HashMap<String, P<ast::Expr>>)> {
-    let mut args = Vec::new();
+    let mut args = vec![];
```
Could changes like this be left out?

alexcrichton commented:
This seems to have gotten pretty significantly more complicated with some unfortunate code duplication in a few places, would it be possible perhaps reduce the number of maps and index pointers in play? One easy optimization would be to just translate all named arguments immediately to numbered indexes, that way there's no branch for each.

xen0n commented:
> One easy optimization would be to just translate all named arguments immediately to numbered indexes, that way there's no branch for each. 

Nice insight! Will try to decrease the complexity this way. Previously I wanted to keep the original machinery intact as much as possible, but the patch have grown beyond my expectation.

xen0n commented:
Addressed the review comments. By applying the named argument optimization I was able to trim down the patch significantly (280 -> 113 net added lines)! I'll squash the review fix commits once everything looks good.

alexcrichton commented on src/libfmt_macros/lib.rs:
```diff
@@ -259,9 +262,41 @@ impl<'a> Parser<'a> {
+                format.precision = CountIsParam(i);
+            }
+            _ => {}
+        }
```
I wonder if it might be best to just remove the `CountIsNextParam` and `ArgumentNext` enum variants? That way this could be handled directly in the `position` and `format` methods

xen0n replied:
I'm afraid we can't, as references to them are already indirectly pulled into existing compiled libraries, thus removing them would mean an ABI breakage. The best we could do is to not use them in newer code and hope that it'll die out some time.

alexcrichton replied:
Oh the `fmt_macros` library is actually entirely unstable so we can change the APIs whenever we want, no need to preserve backwards compatibility

xen0n replied:
So Rust doesn't allow intermixing object code from different compiler versions? I'll try to do that without private enum variants though.

alexcrichton replied:
Yeah we don't need to worry about sharing like that. It should be fine to just drop these enum variants in `fmt_macros`

alexcrichton commented on src/libsyntax_ext/format.rs:
```diff
@@ -131,24 +145,49 @@ fn parse_args(ecx: &mut ExtCtxt, sp: Span, tts: &[ast::TokenTree])
-    /// fatal so we can continue giving errors about this and possibly other
-    /// format strings.
+    fn resolve_name_inplace(&self, p: &mut parse::Piece) {
+        let lookup = |s| *self.names.get(s).unwrap_or(&0);
```
Is this `unwrap_or` branch ever taken? It seems like this may want to be `unwrap()` to indicate a bug

xen0n replied:
It's taken during parse of illegal invocations like in `format_args!("{foo}")` where the piece tries to reference a non-existent named argument, so this is necessary. Maybe I should add a comment regarding this.

alexcrichton replied:
Ah that makes sense, and yeah feel free to add a comment!

alexcrichton commented:
Thanks @xen0n! I'll take a closer look once the RFC is resolved.

alexcrichton commented on src/libsyntax_ext/format.rs:
```diff
@@ -211,84 +244,70 @@ impl<'a, 'b> Context<'a, 'b> {
+                    Count => {
+                        match self.count_positions.entry(arg) {
+                            Entry::Vacant(e) => {
+                                let i = self.count_positions_count;
```
Does this mean that `count_positions_count` an alias for the length of the `count_positions` map?

alexcrichton commented on src/libsyntax_ext/format.rs:
```diff
@@ -211,84 +244,70 @@ impl<'a, 'b> Context<'a, 'b> {
+        for i in 0..args_len {
+            let ref arg_types = self.arg_types[i];
+            let mut arg_offsets = Vec::with_capacity(arg_types.len());
+            for ref offset in arg_types {
```
I think this `ref` can be elided as `arg_types` is already `&Vec<T>` so it yields references when iterated over.

alexcrichton commented on src/libsyntax_ext/format.rs:
```diff
@@ -357,25 +380,34 @@ impl<'a, 'b> Context<'a, 'b> {
+                        parse::ArgumentIs(i) => {
+                            // Map to index in final generated argument array
+                            // in case of multiple types specified
+                            let arg_idx = if self.args.len() > i {
```
This `if` may be best expressed with some form of `<[T]>::get` plus a `match`

alexcrichton commented on src/libsyntax_ext/format.rs:
```diff
@@ -623,10 +650,10 @@ pub fn expand_format_args<'cx>(ecx: &'cx mut ExtCtxt, sp: Span,
                                     -> P<ast::Expr> {
-    let arg_types: Vec<_> = (0..args.len()).map(|_| None).collect();
+    let arg_types: Vec<_> = (0..args.len()).map(|_| Vec::new()).collect();
+    let arg_unique_types: Vec<_> = (0..args.len()).map(|_| Vec::new()).collect();
```
Nowadays this may be best expressed as:

```rust
vec![Vec::new(); args.len()]
```

xen0n replied:
That would require `ArgumentType` to also derive `Clone`. Is this acceptable, as we actually never clone instances of it?

alexcrichton replied:
Ah right, that's fine to leave this as-is then, no need to add `Clone`

alexcrichton commented on src/libsyntax_ext/format.rs:
```diff
@@ -43,17 +44,16 @@ struct Context<'a, 'b:'a> {
-    names: HashMap<String, P<ast::Expr>>,
-    name_types: HashMap<String, ArgumentType>,
-    name_ordering: Vec<String>,
+    /// Placeholder slot numbers indexed by argument.
```
Could you expand this a bit as well? I found this kinda hard to follow, but it looks like if you have a format string like `{0:x} {0:X} {}` this would be something like:

```
[ 
    [0, 1],
    [0],
]
```

where the indexes are pointers into `arg_unique_types`?

alexcrichton commented on src/libsyntax_ext/format.rs:
```diff
@@ -43,17 +44,16 @@ struct Context<'a, 'b:'a> {
+    arg_types: Vec<Vec<usize>>,
+    /// Unique format specs seen for each argument.
+    arg_unique_types: Vec<Vec<ArgumentType>>,
+    /// Map from named arguments to their resolved indices.
```
There's quite a few indexes flying around here, so it may be worth expanding on exactly what "indice" this is containing?

alexcrichton commented on src/libsyntax_ext/format.rs:
```diff
@@ -65,10 +65,27 @@ struct Context<'a, 'b:'a> {
+    /// Relative slot numbers for count arguments.
+    count_positions: HashMap<usize, usize>,
+    /// Number of count slots assigned.
+    count_positions_count: usize,
```
I may be getting confused, but before there were no entries in this structure tracking `count_*` fields. How come these all cropped up now?

xen0n commented:
@alexcrichton Just a quick note that I might need another 2~3 days in completing this; I have another project going at the moment (just done a CyanogenMod port few days ago and already tester feedbacks are pouring at me). I'll ping the issue with commits when I'm done!

bors commented:
:umbrella: The latest upstream changes (presumably #34424) made this pull request unmergeable. Please resolve the merge conflicts.

xen0n commented:
Rebased and addressed most of review comments pending:

* [ ] decision whether `ArgumentType` needs to derive `Clone` only for saner initialization of `Vec<Vec<ArgumentType>>`
* [ ] comment and documentation fixes

bors commented:
:umbrella: The latest upstream changes (presumably #34638) made this pull request unmergeable. Please resolve the merge conflicts.

xen0n commented:
Rebased again and (finally found time to do so!) added comments to facilitate better understanding. Hope this is not too late for 1.10!

alexcrichton commented:
Oops, sorry this fell off the radar! I think this got lost in my inbox by accident...

Ok, everything looks good to me, thanks @xen0n! Could you also squash the commits down? Other than that r=me

xen0n commented:
Squashed all review-generated commits into respective commits and rebased again. Also clarified a few `_ => panic!` cases for better maintainability (no worries if some enum variants got added someday). Should be good to go now! @alexcrichton

alexcrichton commented:
@bors: r+ 51e54e57a45d7b956e3e3390db76ea2651f5d3b3

Thanks again!

bors commented:
:hourglass: Testing commit 51e54e5 with merge db71987...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/9795), [auto-linux-32-opt](https://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/9835), [auto-linux-64-cargotest](https://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/1088), [auto-linux-64-cross-freebsd](https://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/900), [auto-linux-64-debug-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/3061), [auto-linux-64-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/9772), [auto-linux-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/9753), [auto-linux-64-opt-mir](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-mir/builds/1346), [auto-linux-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/1724), [auto-linux-64-x-android-t](https://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/9747), [auto-linux-cross-opt](https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/2938), [auto-linux-musl-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/4903), [auto-mac-32-opt](https://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/9815), [auto-mac-64-opt](https://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/9786), [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/1741), [auto-mac-cross-ios-opt](https://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/911), [auto-win-gnu-32-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/4879), [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/1751), [auto-win-gnu-64-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/4866), [auto-win-msvc-32-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/4068), [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/1096), [auto-win-msvc-64-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/4871), [auto-win-msvc-64-opt-mir](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-mir/builds/1359), [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/1752)
Approved by: alexcrichton
Pushing db71987ee1198be60fef3e361ad1ce59a70cd7f7 to master...
